src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,490 @@
+											RangeTblEntry *rangeTableEntry, Index
+											rteIndex);
+static List * RequiredAttrNumbersForRelation(RangeTblEntry *relationRte,
+											 PlannerRestrictionContext *
+											 plannerRestrictionContext);
+static ConversionCandidates * CreateConversionCandidates(FromExpr *joinTree,
+														 PlannerRestrictionContext *
+														 plannerRestrictionContext,
+														 List *rangeTableList,
+														 Oid resultRelationId);
+static void GetAllUniqueIndexes(Form_pg_index indexForm, List **uniqueIndexes);
+static RangeTableEntryDetails * GetNextRTEToConvertToSubquery(FromExpr *joinTree,
+															  ConversionCandidates *
+															  conversionCandidates,
+															  PlannerRestrictionContext *
+															  plannerRestrictionContext);
+static void GetRangeTableEntriesFromJoinTree(Node *joinNode, List *rangeTableList,
+											 List **joinRangeTableEntries);
+static void RemoveFromConversionCandidates(ConversionCandidates *conversionCandidates, Oid
+										   relationId);
+static bool AllRangeTableEntriesHaveUniqueIndex(List *rangeTableEntryDetailsList);
+
+/*
+ * RecursivelyPlanLocalTableJoins gets a query and the planner
+ * restrictions. As long as the query is not plannable by router planner,
+ * it converts either a local or distributed table to a subquery.
+ */
+void
+RecursivelyPlanLocalTableJoins(Query *query,
+							   RecursivePlanningContext *context)
+{
+	PlannerRestrictionContext *plannerRestrictionContext =
+		context->plannerRestrictionContext;
+
+	List *rangeTableList = NIL;
+	GetRangeTableEntriesFromJoinTree((Node *) query->jointree, query->rtable,
+									 &rangeTableList);
+
+	Oid resultRelationId = InvalidOid;
+	if (IsModifyCommand(query))
+	{
+		resultRelationId = ModifyQueryResultRelationId(query);
+	}
+
+	if (!ShouldConvertLocalTableJoinsToSubqueries(query, rangeTableList, resultRelationId,
+												  plannerRestrictionContext))
+	{
+		return;
+	}
+	ConversionCandidates *conversionCandidates =
+		CreateConversionCandidates(query->jointree, plannerRestrictionContext,
+								   rangeTableList, resultRelationId);
+
+	while (ShouldConvertLocalTableJoinsToSubqueries(query, rangeTableList,
+													resultRelationId,
+													plannerRestrictionContext))
+	{
+		FromExpr *joinTree = query->jointree;
+		RangeTableEntryDetails *rangeTableEntryDetails =
+			GetNextRTEToConvertToSubquery(joinTree, conversionCandidates,
+										  plannerRestrictionContext);
+
+		RangeTblEntry *rangeTableEntry = rangeTableEntryDetails->rangeTableEntry;
+		Oid relId = rangeTableEntryDetails->rangeTableEntry->relid;
+		List *restrictionList = rangeTableEntryDetails->restrictionList;
+		List *requiredAttributeNumbers = rangeTableEntryDetails->requiredAttributeNumbers;
+		ReplaceRTERelationWithRteSubquery(rangeTableEntry, restrictionList,
+										  requiredAttributeNumbers, context);
+		RemoveFromConversionCandidates(conversionCandidates, relId);
```

Comment:
Do we have an assumption that `ReplaceRTERelationWithRteSubquery()` would definitely replace the subqueries? That might not be the case  as `RecursivelyPlanSubquery()`  may skip recursive planning in case  of  correlated queries.

Should we do anything specific for that case?